### PR TITLE
Firestore: Re-write sample code in FirestoreDataConverter docs

### PR DESCRIPTION
**Pull request by dconeybe**
Re-write the sample code for `FirestoreDataConverter`. The old sample had blatant syntax errors and was lacking in its demonstration of the power of type converters.

Fixes: #7671

### CODE REVIEW

**Review by dconeybe — COMMENTED** [review]
Thanks for the improvements! I've just left a few optional comments for you to consider. Otherwise, LGTM.

File: docs-devsite/firestore_.firestoredataconverter.md [hunk 1]
```diff
@@ -98,43 +98,162 @@ toFirestore(modelObject: PartialWithFieldValue<AppModelType>, options: SetOption
...
+    // Firestore values that perform server-side mutations, such as
+    // `arrayRemove()`, `deleteField()`, and `serverTimestamp()`.
+    await setDoc(documentRef, {
+        title: 'My Life',
+        author: 'Foo Bar',
+        lastUpdatedMillis: serverTimestamp()
+    });
+
+    // The TypeScript compiler will fail to compile if the `data` argument to
+    // `setDoc()` is _not_ be compatible with `WithFieldValue<Post>`. This
```
Comment (markarndt) [inline docs-devsite/firestore_.firestoredataconverter.md:212]: Remove "be": "...is _not_ compatible..."

File: docs-devsite/firestore_.firestoredataconverter.md [hunk 2]
```diff
@@ -98,43 +98,162 @@ toFirestore(modelObject: PartialWithFieldValue<AppModelType>, options: SetOption
...
+    }
+
+    // The `data` argument specified to `updateDoc()` is type checked by the
+    // TypeScript compiler to be compatible with `PostDbModel`. Note that
+    // unlike `setDoc()`, whose `data` argument must be compatible with `Post`,
+    // the `data` argument to `updateDoc()` must be compatible with
+    // `PostDbModel`. Similar to `setDoc()`, since the `data` argument is typed
+    // as `WithFieldValue<PostDbModel>` rather than just `PostDbModel`, this
+    // allows properties of the `data` argument to also be those special
+    // Firestore values, like as `arrayRemove()`, `deleteField()`, and
```
Comment (markarndt) [inline docs-devsite/firestore_.firestoredataconverter.md:237]: Remove "as": "...values, like `arrayRemove()`...

File: docs-devsite/firestore_.firestoredataconverter.md [hunk 3]
```diff
@@ -98,43 +98,162 @@ toFirestore(modelObject: PartialWithFieldValue<AppModelType>, options: SetOption
...
+    // `serverTimestamp()`.
+    await updateDoc(documentRef, {
+        'aut.firstName': 'NewFirstName',
+        lut: serverTimestamp()
+    });
+
+    // The TypeScript compiler will fail to compile if the `data` argument to
+    // `updateDoc()` is _not_ be compatible with `WithFieldValue<PostDbModel>`.
```
Comment (markarndt) [inline docs-devsite/firestore_.firestoredataconverter.md:245]: Remove "be": "...is not compatible with..."

File: docs-devsite/firestore_.firestoredataconverter.md (hunk 2 above)
Comment (markarndt) [reply docs-devsite/firestore_.firestoredataconverter.md:237]: Check corresponding docstrings in firebase_lite.

**Review by markarndt — APPROVED** [review]

**Review by egilmorez — APPROVED** [review]
Thanks!

File: docs-devsite/firestore_.firestoredataconverter.md [hunk 4]
```diff
@@ -101,149 +101,159 @@ toFirestore(modelObject: PartialWithFieldValue<AppModelType>, options: SetOption
...
+    await updateDoc(documentRef, { value: 999 });
+    const snapshot2 = await getDoc(documentRef);
+    assertEqual(snapshot2.data(), 999);
 }
 
 ```
 Advanced Example
 
 ```typescript
+// The Post class is a model that is used by our application.
```
[the remark this answers is not in the document]
Comment (MarkDuckworth) [reply docs-devsite/firestore_.firestoredataconverter.md:131]: Thanks!

File: docs-devsite/firestore_.firestoredataconverter.md [hunk 5]
```diff
@@ -101,149 +101,159 @@ toFirestore(modelObject: PartialWithFieldValue<AppModelType>, options: SetOption
...
-    return new Post(data.ttl, author, data.lut.toMillis());
-  },
-
-  _autFromAuthor(
-    author: string | FieldValue
-  ): { firstName: string; lastName: string } | FieldValue {
-    if (typeof author !== 'string') {
-      // `author` is a FieldValue, so just return it.
-      return author;
+// The `PostConverter` implements `FirestoreDataConverter<AppModelType, DbModelType>`
```
[the remark this answers is not in the document]
Comment (MarkDuckworth) [reply docs-devsite/firestore_.firestoredataconverter.md:155]: Agreed. Thanks.

File: docs-devsite/firestore_.firestoredataconverter.md (hunk 1 above)
Comment (MarkDuckworth) [reply docs-devsite/firestore_.firestoredataconverter.md:212]: Thanks

File: docs-devsite/firestore_.firestoredataconverter.md (hunk 2 above)
Comment (MarkDuckworth) [reply docs-devsite/firestore_.firestoredataconverter.md:237]: Thanks.

File: docs-devsite/firestore_.firestoredataconverter.md (hunk 3 above)
Comment (MarkDuckworth) [reply docs-devsite/firestore_.firestoredataconverter.md:245]: TY

**Pull request merged**